File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if not head.next:
            return None
            
        reverse = self.reverseList(head)

        curr = reverse

        i = 1
        if i == n:
            curr = curr.next
            return self.reverseList(curr)

        while i < n - 1:
            curr = curr.next
            i += 1

        if curr.next:
            curr.next = curr.next.next
        
        return self.reverseList(reverse)
    
    def reverseList(self, head):
        prev = None

        while head:
            head.next, prev, head = prev, head, head.next

        return prev

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-4.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> oldToNew = new HashMap<>();

        Node curr = head;
        while (curr != null) {
            oldToNew.put(curr, new Node(curr.val));
            curr = curr.next;
        }

        curr = head;
        Node res = oldToNew.get(curr);
        Node dummy = new Node(0);
        dummy.next = res;
        
        while (curr != null) {
            res.next = oldToNew.get(curr.next);
            res.random = oldToNew.get(curr.random);

            res = res.next;
            curr = curr.next;
        }

        return dummy.next;
        
    }
}
